tools/serverpod_cli/lib/src/downloads/resource_manager.dart

```diff
@@ -103,33 +104,72 @@ class ResourceManager {
     return templateDirectory.existsSync();
   }
 
+  Future<bool> verifyAndInstallTemplatesIfNeeded(String templateType) async {
+    if (!isTemplatesInstalled) {
+      if (!productionMode) {
+        return true;
+      }
+      try {
+        await installTemplates();
+      } catch (e) {
+        log.error('Failed to download templates: $e');
+        return false;
+      }
+    }
+
+    return await TemplateChecksumManager.verifyWithRetry(
+      templateDirectory,
+      templateType: templateType,
+      redownloadCallback: () async {
+        await installTemplates();
+      },
+    );
+  }
+
   Future<void> installTemplates() async {
     log.info('Downloading templates for version $templateVersion');
     if (!versionedDir.existsSync()) versionedDir.createSync(recursive: true);
 
-    var response = await http.get(Uri.parse(packageDownloadUrl));
-    var data = response.bodyBytes;
-
-    // var outFile = File(p.join(versionedDir.path, 'serverpod_templates.tar.gz'));
-    // outFile.writeAsBytesSync(data);
-
-    // Const constructor was introduced in version 4.0.0 of archive package.
-    // At the time this was written, that version was 2 days old and we don't
-    // want to force a constraint on the package for this.
-    // ignore: prefer_const_constructors
-    var unzipped = GZipDecoder().decodeBytes(data);
-    var archive = TarDecoder().decodeBytes(unzipped);
-
-    for (var file in archive) {
-      var outFileName = p.join(templateDirectory.path, file.name);
-      if (file.isFile) {
-        var outFile = File(outFileName);
-        outFile = await outFile.create(recursive: true);
-        await outFile.writeAsBytes(file.content);
-      } else {
-        await Directory(outFileName).create(recursive: true);
+    var tempDir = Directory(p.join(
+        versionedDir.path, 'temp_${DateTime.now().millisecondsSinceEpoch}'));
+    tempDir.createSync(recursive: true);
```

Comment:
```suggestion
    var tempDir = Directory.systemTemp.createTempSync();
```
this will allow the OS to cleanup if we fail